CosNativeFileSystemStore.java

```diff
@@ -0,0 +1,768 @@
+      LOG.debug("Store file successfully. COS key: [{}], ETag: [{}], "
+          + "MD5: [{}].", key, putObjectResult.getETag(), new String(md5Hash));
+    } catch (Exception e) {
+      String errMsg = String.format("Store file failed. COS key: [%s], "
+          + "exception: [%s]", key, e.toString());
+      LOG.error(errMsg);
+      handleException(new Exception(errMsg), key);
+    }
+  }
+
+  /**
+   * Store a local file into COS.
+   *
+   * @param key     COS object key.
+   * @param file    The local file to be uploaded.
+   * @param md5Hash The MD5 value of the file to be uploaded.
+   * @throws IOException Upload the file failed.
+   */
+  @Override
+  public void storeFile(String key, File file, byte[] md5Hash)
+      throws IOException {
+    LOG.info("Store file from local path: [{}]. file length: [{}] COS key: " +
+            "[{}] MD5: [{}].", file.getCanonicalPath(), file.length(), key,
+        new String(md5Hash));
+    storeFileWithRetry(key, new BufferedInputStream(new FileInputStream(file)),
+        md5Hash, file.length());
+  }
+
+  /**
+   * Store a file into COS from the specified input stream.
+   *
+   * @param key           COS object key.
+   * @param inputStream   The Input stream to be uploaded.
+   * @param md5Hash       The MD5 value of the content to be uploaded.
+   * @param contentLength Length of uploaded content.
+   * @throws IOException Upload the file failed.
+   */
+  @Override
+  public void storeFile(
+      String key,
+      InputStream inputStream,
+      byte[] md5Hash,
+      long contentLength) throws IOException {
+    LOG.info("Store file from input stream. COS key: [{}], "
+        + "length: [{}], MD5: [{}].", key, contentLength, md5Hash);
+    storeFileWithRetry(key, inputStream, md5Hash, contentLength);
+  }
+
+  // For cos, storeEmptyFile means creating a directory
+  @Override
+  public void storeEmptyFile(String key) throws IOException {
+    if (!key.endsWith(CosNFileSystem.PATH_DELIMITER)) {
+      key = key + CosNFileSystem.PATH_DELIMITER;
+    }
+
+    ObjectMetadata objectMetadata = new ObjectMetadata();
+    objectMetadata.setContentLength(0);
+    InputStream input = new ByteArrayInputStream(new byte[0]);
+    PutObjectRequest putObjectRequest =
+        new PutObjectRequest(bucketName, key, input, objectMetadata);
+    try {
+      PutObjectResult putObjectResult =
+          (PutObjectResult) callCOSClientWithRetry(putObjectRequest);
+      LOG.debug("Store empty file successfully. COS key: [{}], ETag: [{}].",
+          key, putObjectResult.getETag());
+    } catch (Exception e) {
+      String errMsg = String.format("Store empty file failed. "
+          + "COS key: [%s], exception: [%s]", key, e.toString());
+      LOG.error(errMsg);
+      handleException(new Exception(errMsg), key);
+    }
+  }
+
+  public PartETag uploadPart(File file, String key, String uploadId,
+      int partNum) throws IOException {
+    InputStream inputStream = new FileInputStream(file);
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

**Problem**
This line of code might contain a resource leak. Resource leaks can cause your system to slow down or crash.

**Fix**
Consider closing the following resource: *inputStream*. Currently, there are execution paths that do not contain closure statements, e.g., when exception is thrown by UploadPartRequest constructor. Either a) close *inputStream* in a try-finally block or b) close the resource by declaring *inputStream* in a try-with-resources block.

**More info**
[View resource management guidelines at oracle.com](https://www.oracle.com/technetwork/articles/java/trywithresources-401775.html) (external link).